libgalois/src/analytics/louvain_clustering/louvain_clustering.cpp

```diff
@@ -309,11 +314,67 @@ struct LouvainClusteringImplementation
     return prev_mod;
   }
 
+  /// Cache the current community ids for the next phase
+  void UpdatePrevClusterIdsFromCurrIds(
+      katana::LargeArray<CommunityIdTy>* previous_cluster_ids,
+      const Graph& graph_curr, const CommunityIdTy num_prev_clusters,
+      bool is_first_iter, bool is_vf_enabled) {
+    if (!is_vf_enabled && is_first_iter) {
+      KATANA_LOG_DEBUG_ASSERT(num_prev_clusters == graph_curr.num_nodes());
+      katana::do_all(katana::iterate(graph_curr), [&](GNode n) {
+        (*previous_cluster_ids)[n] =
+            graph_curr.template GetData<CurrentCommunityId>(n);
+      });
+    } else {
+      katana::do_all(
+          katana::iterate(CommunityIdTy{0}, num_prev_clusters), [&](GNode n) {
+            CommunityIdTy& n_previous_cluster = (*previous_cluster_ids)[n];
+            if (n_previous_cluster != Base::UNASSIGNED) {
+              KATANA_LOG_DEBUG_ASSERT(
+                  n_previous_cluster < graph_curr.num_nodes());
+              n_previous_cluster =
+                  graph_curr.template GetData<CurrentCommunityId>(
+                      n_previous_cluster);
+            }
+          });
+    }
+  }
+
+  /// Choose and run louvain algorithm
+  katana::Result<void> RunAlgorithm(
+      LouvainClusteringPlan plan, katana::PropertyGraph* pfg_curr,
+      ModularityTy* curr_mod, uint32_t iter) {
+    switch (plan.algorithm()) {
+    case LouvainClusteringPlan::kDoAll: {
+      auto curr_mod_result = LouvainWithoutLockingDoAll(
+          pfg_curr, *curr_mod, plan.modularity_threshold_per_round(), &iter);
+      if (!curr_mod_result) {
+        return curr_mod_result.error();
+      }
+      *curr_mod = curr_mod_result.value();
+      break;
+    }
+    case LouvainClusteringPlan::kDeterministic: {
+      auto curr_mod_result = LouvainDeterministic(
+          pfg_curr, *curr_mod, plan.modularity_threshold_per_round(), &iter);
+      if (!curr_mod_result) {
+        return curr_mod_result.error();
+      }
+      *curr_mod = curr_mod_result.value();
+      break;
+    }
+    default:
+      return katana::ErrorCode::InvalidArgument;
+    }
+    return katana::ResultSuccess();
+  }
+
 public:
   katana::Result<void> LouvainClustering(
       katana::PropertyGraph* pfg, const std::string& edge_weight_property_name,
       const std::vector<std::string>& temp_node_property_names,
-      katana::LargeArray<uint64_t>* clusters_orig, LouvainClusteringPlan plan) {
+      katana::LargeArray<CommunityIdTy>* previous_cluster_ids,
```

Comment:
I just wanted to emphasize that the clusters_orig was meant to keep track of the community ids of the nodes in the original graph.